### ISSUE TITLE
Simplification Removes Road Intersections From OSM Tiles

### ISSUE THREAD

**Issue by backspaces** [mapbox/tippecanoe#759]
MapBox Support and I have been working on this issue:
> Mapbox Vector Tiles appear to lack accurate intersection nodes. Is this a bug?

Basically, the default line simplification removes intersections from OSM data. Here street intersections are any nodes with 3 or more lines using the node.

Support suggest I try turning off simplification. Indeed this resulted in correct intersection data but much larger vector tiles. Here are before/after examples of tile 14/3370/6451 extracted from OSM data using the bounding box of 10/210/403 and run through tippecanoe with default setting, and then with no simplification:
> tippecanoe -z14 -f -e tiles10 santafe10.json
> tippecanoe -z14 -f --no-line-simplification -e tiles10ns santafe10.json

The datasets for the MBStudio were created using the two resulting 14/3370/6451 tiles and using MB's vt2geojson
> vt2geojson tiles10//14/3370/6451.pbf > santafe14vt.json

The red circles show the intersection nodes being simplified out of the tile, while the rectangles show the increase in node count.

![Snap 05 10 19-16 53 27](https://user-images.githubusercontent.com/124366/57589257-c2b95b80-74de-11e9-9cb0-13cb3da3e059.jpg)

This issue is either a documentation request on how to use tippecanoe as-is to achieve the desired result, or a feature request for control over simplification to insure certain nodes are not simplified away.

If the former, I'd guess the documentation would suggest insuring intersections be the first or last nodes in a LineString. This may be a solution if the simplification algorithm works on individual LineStrings, and simplification does not join LineStrings during simplification.

If the latter, I'd guess it would require somehow specifying the nodes to be preserved, either algorithmically (node-line count, say), or by a list of node coordinates to be preserved. Or if simplification does join LineStrings, an option for not doing so unless the string is invisible at the given Z.

Whew! Hard to explain all this. I hope I've been clear. Please let me know if I can provide more info or clarify the issue. I realize there is a tension between "draw" and "data" tile usage, street intersections being unusual in that, even tho the street tiles "draw" well, the underlying "data" isn't useful for computing traffic flow.

**Comment by ericfischer** [comment]
I think the best way to do this currently would be to represent the OSM ways as MultiLineStrings instead of LineStrings, with each block face represented by a separate LineString within the MultiLineString, since each of these should be simplified separately. (If you use `--coalesce`, there is a second simplification pass that still might remove these shared nodes.)

If that doesn't work for you, I could look into the possibility of adding an option to prevent simplifying away shared nodes.

**Comment by ericfischer** [comment]
Actually I am wrong and it is not sufficient to use a MultiLineString. I'll look into adding explicit support for intersections.

**Comment by backspaces** [comment]
Yay! And thanks.

**Comment by ericfischer** [comment]
Could you try `--no-simplification-of-shared-nodes` in https://github.com/mapbox/tippecanoe/pull/760 and see if it does what you are looking for?

**Comment by backspaces** [comment]
I tried cloning the branch and building it per the instructions but it died when I invoked `make -j`
> xcrun: error: invalid active developer path (/Library/Developer/CommandLineTools), missing xcrun at: /Library/Developer/CommandLineTools/usr/bin/xcrun

Is there an easier way for me to do this (mac os x mojave) than a full  build? Or developer instructions for mac os x? Sorry for being so noob!

**Comment by ericfischer** [comment]
Sorry, I don't have any better suggestion than building from source. I've never seen that `xcrun` error, but I think this means you need to install the XCode command line tools with `xcode-select --install`.

**Comment by backspaces** [comment]
TLDR: When running with the new binary, the resulting json from the vector tiles won't load into MB Studio and have odd repeated coordinates for each feature:
```
      "coordinates": [
        [
          {
            "x": -44,
            "y": 211
          },
          {
            "x": 0,
            "y": 195
          },
          {
            "x": 28,
            "y": 186
          }
        ]
      ]
```

Details:
OK, finally got xcode fixed, sigh. `make -j` ran with no errors. I did not run `make install`, not wanting to install the test in /usr/local.

I then repeated prior tests. 

First, create an osm and geojson for the zoom 10 for santa fe nm:
Use osm request for bbox of tile of zoom 10 for santa fe:
```
curl 'https://overpass-api.de/api/interpreter?data=%5Bout%3Ajson%5D%5Btimeout%3A180%5D%5Bbbox%3A35.4606699514953%2C-106.171875%2C35.74651225991851%2C-105.8203125%5D%3B%0Away%5Bhighway%5D%3B%0A(._%3B%3E%3B)%3B%0Aout%3B' > test.osm
```
convert osm to geojson
```
osmtogeojson < test.osm > test.json
```
Create the tile set:
```
~/dl/tippecanoe-preserve-shared-nodes/tippecanoe -z14 -f --no-simplification-of-shared-nodes -e tiles10test test.json
    For layer 0, using name "test"
    Warning: Can't represent non-numeric feature ID "way/319672018"
    15340 features, 883929 bytes of geometry, 2292 bytes of separate metadata, 314367 bytes of string pool
    99.9%  14/3372/6449
```
Tile set sizes make sense:
```
        5.2M	tiles10/        default, simplification
        5.7M	tiles10ns/      no simplification
        5.4M	tiles10test/    no shared simplification
```
Convert tile 14/3370/6451.pbf to json for upload to MB Studio:
```
vt2geojson tiles10test//14/3370/6451.pbf > test14.json
```
The three file sizes make sense
```
     13M May 14 12:32 test.osm
     11M May 14 12:37 test.json
    1.6M May 14 12:46 test14.json
```
Finally, the resulting test14.json appears to be malformed. Here are the test files:
[testfiles.zip](https://github.com/mapbox/tippecanoe/files/3179499/testfiles.zip)

**Comment by ericfischer** [comment]
Thanks. I'll take a look and see if I can figure out what is going wrong.

**Comment by ericfischer** [comment]
That extra `coordinates` array with integer coordinates is something that your copy of `vt2geojson` is producing, not anything that has to do with the version of Tippecanoe that you are using. It is the raw unprojected coordinates from the tile. It is not happening with the copy of `vt2geojson` that I just installed myself with `npm install -g @mapbox/vt2geojson`

**Comment by ericfischer** [comment]
I was also able to upload your `test14.json` file through Mapbox Studio without any error messages.

**Comment by backspaces** [comment]
Great! So it was just that my vt2geojson was out of date? Sorry!  

When I upgraded vt2geojson@1.1.0 -> @mapbox/vt2geojson@1.1.5 everything worked as you found and indeed I could upload test14.json and it looks to have intersections intact, yay!

This means --no-simplification-of-shared-nodes preserves intersections and we can close this.
